multidict/_multidict_c.c

```diff
@@ -0,0 +1,1587 @@
+        _default != NULL)
+    {
+        PyErr_Clear();
+        Py_INCREF(_default);
+        return _default;
+    }
+
+    return list;
+}
+
+static PyObject *
+multidict_getone(MultiDictObject *self, PyObject *args, PyObject *kwds)
+{
+    PyObject *key      = NULL,
+             *_default = NULL;
+
+    static char *getone_keywords[] = {"key", "default", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O|O:getone",
+                                     getone_keywords, &key, &_default))
+    {
+        return NULL;
+    }
+
+    return _multidict_getone(self, key, _default);
+}
+
+static PyObject *
+multidict_keys(MultiDictObject *self)
+{
+    return multidict_keysview_new((PyObject*)self);
+}
+
+static PyObject *
+multidict_items(MultiDictObject *self)
+{
+    return multidict_itemsview_new((PyObject*)self);
+}
+
+static PyObject *
+multidict_values(MultiDictObject *self)
+{
+    return multidict_valuesview_new((PyObject*)self);
+}
+
+static PyObject *
+multidict_reduce(MultiDictObject *self)
+{
+    PyObject *items      = NULL,
+             *items_list = NULL,
+             *args       = NULL,
+             *result     = NULL;
+
+    items = multidict_items(self);
+    if (items == NULL) {
+        goto ret;
+    }
+
+    items_list = PySequence_List(items);
+    if (items_list == NULL) {
+        goto ret;
+    }
+
+    args = PyTuple_Pack(1, items_list);
+    if (args == NULL) {
+        goto ret;
+    }
+
+    result = PyTuple_Pack(2, Py_TYPE(self), args);
+
+ret:
+    Py_XDECREF(args);
+    Py_XDECREF(items_list);
+    Py_XDECREF(items);
+
+    return result;
+}
+
+static PyObject *
+multidict_repr(PyObject *self)
+{
+    return PyObject_CallFunctionObjArgs(
+        repr_func, self, NULL);
+}
+
+static Py_ssize_t
+multidict_mp_len(MultiDictObject *self)
+{
+    return pair_list_len(self->impl);
+}
+
+static PyObject *
+multidict_mp_subscript(MultiDictObject *self, PyObject *key)
+{
+    return _multidict_getone(self, key, NULL);
+}
+
+static int
+multidict_mp_as_subscript(MultiDictObject *self, PyObject *key, PyObject *val)
+{
+    if (val == NULL) {
+        return pair_list_del(self->impl, key);
+    } else {
+        return pair_list_replace(self->impl, key, val);
+    }
+}
+
+static int
+multidict_sq_contains(MultiDictObject *self, PyObject *key)
+{
+    return pair_list_contains(self->impl, key);
+}
+
+static PyObject *
+multidict_tp_iter(MultiDictObject *self)
+{
+    return PyObject_GetIter(multidict_keysview_new((PyObject*)self));
+}
+
+static PyObject *
+multidict_tp_richcompare(PyObject *self, PyObject *other, int op)
+{
+    // TODO: refactoring me with love
+
+    int cmp = 0;
+
+    if (op != Py_EQ && op != Py_NE) {
+        Py_RETURN_NOTIMPLEMENTED;
+    }
+
+    if (MultiDict_CheckExact(other) || CIMultiDict_CheckExact(other)) {
+        cmp = _multidict_eq(
+            (MultiDictObject*)self,
+            (MultiDictObject*)other
+        );
+        if (cmp < 0) {
+            // TODO: set exception
```

Comment:
What about this?

Reply:
`cmp < 0` if `_multidict_eq()` returns -1.
The Python exception is set here by `PyObject_RichCompareBool()`